### PR TITLE
Update to Yarn 3

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Any Dependency Changes?  | Yes, Yarn (committed dev dependency)
| License                  | MIT

Yarn 3 isn't stable yet, but I want to verify that it doesn't introduce any unexpected regression in our workflow.

~~It currently breaks our release tool, but it's a known breaking change.~~

What works so far:
- [x] `yarn install` (also with the `condition:` plugin)
- [x] `yarn install` works on CI (https://github.com/yarnpkg/berry/pull/2530)
- [x] `yarn release-tool version`
- [x] `yarn release-tool publish`

You can see the breaking changes list at https://github.com/yarnpkg/berry/blob/master/CHANGELOG.md#breaking-changes. We are affected by:
- Node.js 10 has been dropped
- Yarn runs with `--immutable` by default on CI
There are workarounds/changes for both of them.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13140"></a>

### CODE REVIEW

**Comment by merceyz** [comment]
Would you be willing to change this PR to use the CLI from https://github.com/yarnpkg/berry/pull/2161 and enable PnP for the build step again? I updated the PR over the weekend and added tests based on your comments from last time and fixed some issues that caused it to crash when running `yarn gulp` in this repo, it looks like it works but would be nice to have the CI agree.

**Comment by nicolo-ribaudo** [comment]
Ok, sure!

**Comment by nicolo-ribaudo** [comment]
When using PnP the generated `@babel/standalone` bundle is broken, but it might be caused by something else (maybe Rollup).

**Comment by merceyz** [comment]
Yeah I saw that, looked into the bundle but nothing stood out, I see the methods are missing but not sure why

**Comment by merceyz** [comment]
These two imports are merged into one https://github.com/facebook/regenerator/blob/dbbddd9bd8c099ed1cdfcecdd077b8b4e7f43042/packages/transform/src/leap.js#L10-L11 so it's trying to get `inherits` from `./util` which doesn't exist

**Comment by merceyz** [comment]
Got some time to look into it and it looks like a bug on our end, I'll fix it

**Comment by nicolo-ribaudo** [comment]
Oh thanks!

**Comment by nicolo-ribaudo** [comment]
Yarn 3 has been released! I removed the commit to test https://github.com/yarnpkg/berry/pull/2161 since that PR has been deferred to a later version. I'll update a new branch PR if we'll have to test ESM support before that it's merged.

**Comment by nicolo-ribaudo** [comment]
@arcanis Was the `--skip-builds` option removed? It's not mentioned in the changelog :thinking:

**Comment by arcanis** [comment]
It got replaced by `--mode=skip-build`, I'll add it to the changelog 👍

**Comment by nicolo-ribaudo** [comment]
The CI failure looks related

**Comment by arcanis** [comment]
Running the following locally seems to work for me - could it just be that your trunk has legit breaking changes that would break using a version of `@babel/types` at the same time as the trunk one?

```
rm -rf /tmp/jest
git clone --depth=1 https://github.com/facebook/jest /tmp/jest 
cd /tmp/jest
node -e "
  var pkg = require('./package.json');
  pkg.resolutions || (pkg.resolutions = {});
  pkg.resolutions['@types/babel__traverse/@babel/types'] = 'latest';
  fs.writeFileSync('./package.json', JSON.stringify(pkg, null, 2));
"
yarn install
yarn build
```

**Comment by nicolo-ribaudo** [comment]
Yeah, I think that "related" is more something on our side uncovered by an innocuous change in Yarn.

@larixer I don't remember, I'll revert it back :sweat_smile:

**Comment by nicolo-ribaudo** [comment]
Ok, it looks like I used npm to install verdaccio because Yarn 3 throws:
```
+ yarn global add verdaccio-memory@~9.7.2
Usage Error: Couldn't find the node_modules state file - running an install might help (findPackageLocation)
```

**Comment by merceyz** [comment]
It wants you to run an install before executing scripts which is what it thinks you want to do since `global` isn't a Yarn command anymore

**Comment by larixer** [comment]
@nicolo-ribaudo You should use `YARN_IGNORE_PATH=1 yarn global add verdaccio-memory@~9.7.2` to install verdaccio via "system-wide" yarn binary

File: scripts/integration-tests/e2e-jest.sh [hunk 1]
```diff
@@ -48,6 +48,13 @@ if [ "$BABEL_8_BREAKING" = true ] ; then
   # Jest depends on @types/babel__traverse for Babel 7, and they contain the removed Noop node
   sed -i 's/t.Noop/any/g' node_modules/@types/babel__traverse/index.d.ts
   sed -i 's/t.Noop/any/g' node_modules/@types/babel__traverse/ts4.1/index.d.ts
+
+  node -e "
+    var pkg = require('./package.json');
+    pkg.resolutions || (pkg.resolutions = {});
+    pkg.resolutions['@types/babel__traverse/@babel/types'] = 'latest';
+    fs.writeFileSync('./package.json', JSON.stringify(pkg, null, 2));
+  "
```
Comment (nicolo-ribaudo) [inline scripts/integration-tests/e2e-jest.sh:57]:
```suggestion
```

**Comment by JLHwung** [comment]
The coverage test is stale after 4h: https://github.com/babel/babel/pull/13140/checks?check_run_id=3546719297 I have terminated the workflow.

**Comment by arcanis** [comment]
I think this issue is caused by Node 16.9 and Jest; cf https://github.com/nodejs/node/issues/40030

**Comment by nicolo-ribaudo** [comment]
I added a `console.log(this.tag)` statement at https://github.com/nicolo-ribaudo/yarn-plugin-babel-release-tool/blob/0b7b5e65f2b7c9db6cf2c83c96e69840de96776b/sources/commands/publish.ts#L49, and when I run `yarn release-tool publish --yes --tag-version-prefix="version-e2e-test-"` it logs `undefined` (this is why the test on CircleCI is failing).

What's the correct way to set the default value for an optional flag (`--tag`) in Yarn 3?

**Comment by nicolo-ribaudo** [comment]
Oh I need to use the clipanion 3 interface.

**Comment by merceyz** [comment]
We have a compat layer that is supposed to handle that 🤔

File: package.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
     "test:runtime:bundlers": "node test/runtime-integration/bundlers.cjs",
     "test:runtime:node": "node test/runtime-integration/node.cjs"
   },
+  "packageManager": "yarn@3.0.0",
```
Comment (nicolo-ribaudo) [inline package.json:19]: Q: Is this for `corepack`?

Comment (merceyz) [reply package.json]: Yes, `corepack` will launch `yarn@3.0.0` which will check `yarnPath` and use that one if it doesn't match

**Review by merceyz — COMMENTED** [review]
CI is passing but `@yarnpkg/plugin-constraints` should be updated as well

**Comment by nicolo-ribaudo** [comment]
Greeeen :partying_face:

**Comment by nicolo-ribaudo** [comment]
After merging this I'll do something like https://github.com/yarnpkg/berry/pull/3431

**Review by hzoo — APPROVED** [review]
nice work, verifying is much easier than testing ci..

**Review by JLHwung — APPROVED** [review]

**Comment by merceyz** [comment]
The issue in https://github.com/babel/babel/pull/13140#issuecomment-845823492 has been fixed and the ESM PR has been updated

**Pull request merged**